### PR TITLE
Change `array_agg` to return `null` on no input rather than empty list

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

As @findepi pointed out in https://github.com/apache/datafusion/issues/11274#issuecomment-2211364784 that most of the aggregate function does not return non-null result if no row qualified. I double check the result in Postgres and Duckdb and find out they does not return empty list for `array_agg`. I think we can follow the behaviour as they did.

I also hope this can make dealing with nullability simpler
1. Check the nullability of each (aggregate) function
2. Make sure nullable is helpful for logical optimizer.

Closes #.

## Are there any user-facing changes?

The result of array agg is changed

### CODE REVIEW

File: datafusion/sqllogictest/test_files/aggregate.slt [hunk 1]
```diff
@@ -2744,28 +2735,84 @@ SELECT ARRAY_AGG([1])
...
 ----
-[[1]]
+
+query ?I
+select array_agg(a), count(1) from t where a > 3 group by a;
+----
+
+# TODO: Expect NULL, got empty list
+query ?
+select array_agg(distinct a) from t where a > 3;
```
Comment (jayzhan211) [inline datafusion/sqllogictest/test_files/aggregate.slt:2773]: This is the only query that has unexpected result.

File: datafusion/sqllogictest/test_files/aggregate.slt [hunk 2]
```diff
@@ -1753,31 +1753,12 @@ NULL 4 29 1.260869565217 123 -117 23
...
-  SELECT array_agg(distinct c2) as arr, count(1) as dummy FROM aggregate_test_100
-) data
-  CROSS JOIN indices
-ORDER BY 1
-----
-1 5 100
-2 5 100
-3 5 100
-4 5 100
-5 5 100
```
Comment (findepi) [inline datafusion/sqllogictest/test_files/aggregate.slt:1780]: why removed?

File: datafusion/sqllogictest/test_files/aggregate.slt (hunk 1 above)
Comment (findepi) [reply datafusion/sqllogictest/test_files/aggregate.slt:2773]: Thanks @jayzhan211 for calling this out!

Given there are no rows with `a > 3`, the aggregation input relation `from t where a > 3` is empty, so this should produce same result as array_agg from empty table, ie `NULL`. currently it shows `[]` as a result.

File: datafusion/sqllogictest/test_files/aggregate.slt [hunk 3]
```diff
@@ -2744,28 +2735,98 @@ SELECT ARRAY_AGG([1])
...
+----
+
+query ?I
+select array_agg(a order by a), count(1) from t where a > 3 group by a;
+----
+
+statement ok
+drop table t;
+
+# test with no values
```
Comment (findepi) [inline datafusion/sqllogictest/test_files/aggregate.slt:2808]: add `array_agg(distinct` case on empty table

File: datafusion/sqllogictest/test_files/aggregate.slt (hunk 2 above)
Comment (jayzhan211) [reply datafusion/sqllogictest/test_files/aggregate.slt:1780]: I rewrite it to the simpler one!

File: datafusion/sqllogictest/test_files/aggregate.slt (hunk 1 above)
Comment (alamb) [reply datafusion/sqllogictest/test_files/aggregate.slt:2773]: Maybe it is related to a multi-stage grouping (where the first phase outputs a NULL and then the second phase gets that NULL and interprets it as `[]` 🤔

Comment (alamb) [reply datafusion/sqllogictest/test_files/aggregate.slt:2774]: The results look good now

**Review by alamb — APPROVED** [review]
Looks good to me @jayzhan211 . Thank you for this PR

Also, thank you @findepi for the reviews 🙏 

I double checked the answers in postgres and I believe after this PR DataFusion has consistent behavior

```sql
postgres=# create table t(a int, b float, c bigint);
ERROR:  relation "t" already exists
postgres=# drop table t;
DROP TABLE
postgres=# create table t(a int, b float, c bigint);
CREATE TABLE
postgres=# insert into t values (1, 1.2, 2);
INSERT 0 1
postgres=# select array_agg(a) from t where a > 2;
 array_agg
-----------

(1 row)

postgres=# select array_agg(b) from t where b > 3.1;
 array_agg
-----------

(1 row)

postgres=# select array_agg(c), count(1) from t where c > 3;
 array_agg | count
-----------+-------
           |     0
(1 row)

postgres=# select array_agg(distinct a) from t where a > 3;
 array_agg
-----------

(1 row)
```

**Comment by jayzhan211** [comment]
Thanks @alamb @findepi

**Pull request merged**